draft-ietf-taps-interface.md

```diff
@@ -1780,37 +1772,179 @@ that the other endpoint is done sending on a connection.
+and since many of these protocols require message boundaries to function,
+each application layer protocol has defined its own framing.
+
+While many protocols can be represented as Message Framers, for the 
+purposes of the Transport Services interface these are ways for applications
+or application frameworks to define their own Message parsing to be
+included within a Connection's Protocol Stack. As an example, TLS can
+serve the purpose of framing data over TCP, but is exposed as a protocol
+natively supported by the Transport Services interface.
 
-The Receive Event is intended to be fired once per application-layer Message
-sent by the remote endpoint; i.e., it is a desired property of this interface
-that a Send at one end of a Connection maps to exactly one Receive on the
-other end. This is possible with Protocol Stacks that provide
-message boundary preservation, but is not the case over Protocol Stacks that
-provide a simple octet stream transport.
+Most Message Framer protocols fall into one of two categories:
+- Header-prefixed record protocols, such as a basic Type-Length-Value (TLV) structure
+- Delimeter-separated protocols, such as HTTP/1.1.
 
-For preserving message boundaries over stream transports, this interface
-provides receiver-side de-framing. This facility is based on the observation
-that, since many of our current application protocols evolved over TCP, which
-does not provide message boundary preservation, and since many of these protocols
-require message boundaries to function, each application layer protocol has
-defined its own framing. A Deframer allows an application to push this
-de-framing down into the interface, in order to transform an octet stream into
-a sequence of Messages.
+Note that while Message Framers add the most value when placed above
+a protocol that otherwise does not preserve message boundaries, they can
+also be used with datagram- or message-based protocols. In these cases,
+they add an additional transformation to further encode or encapsulate,
+and can potentially support packing multiple application-layer Messages
+into individual transport datagrams.
 
-Concretely, receiver-side de-framing allows a caller to provide the interface
-with a function that takes an octet stream, as provided by the underlying
-Protocol Stack, reads and returns a single Message of an appropriate type for
-the application and platform, and leaves the octet stream at the start of the
-next Message to deframe. It consists of a Deframer Object with a single Action,
-Deframe. Since the Deframer depends on the protocol used at the application
-layer, it is bound to the Preconnection during the pre-establishment phase:
+## Defining Message Framers
 
+Applications can define a Message Framer by creating an object to represent a unique
+definition of a framing protocol. Event handlers for inbound and outbound data
+are registered on this object.
+
+~~~
+framer := NewMessageFramer()
+framer.setEventHandlers()
```

Comment:
You need something to add to the connection. The MessageFramer is the object that the application creates to do this.

Reply:
Oh I finally I think I understand where my confusion comes from. I assumed that the application was free to create their own framer class that had to expose certain functions for interaction with the connection, so I was looking for that interface. Sorry, my bad.

The place where the application now implements the behavior of framer is in the eventHandlers it sets, is that correct?

Reply:
Thank you Brian, thats exactly what I couldn't quite put into words. 

The previous iteration of framers made it very simple for applications that had no interest in implementing them to grab a framer implementation someone else provided and use it. All the calls and callbacks between the implementation and the TAPS system stayed the same, independent of whether or not a framer was used. I think its important to keep the part of the interface application developers use simple while adding complexity in the part someone that implements a framer would use, because at least to me these are two different groups of people.

Reply:
Agreed. The code here is just how you implement a custom framer. The calls to actually add it to a connection (the common action) is below.

Reply:
Of course, my bad. I totally missed that part, that makes things a lot clearer and I now understand the advantage of doing it this way around, thanks.

On a different note, why is the framer exposing a setEventHandlers call while neither the preconnection nor the connection expose something similar even though they issue events as well?